src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp

```diff
@@ -154,10 +157,75 @@ std::unordered_set<irep_idt> gen_loop_locals_set(
   return loop_locals;
 }
 
-assignst dfcc_infer_loop_assigns(
+/// Append all identifiers in `expr` into `identifiers`.
+static void
+append_ids_in_expr(const exprt &expr, std::unordered_set<irep_idt> &identifiers)
+{
+  auto ids = find_symbol_identifiers(expr);
+  identifiers.insert(ids.begin(), ids.end());
+}
+
+/// Find all identifiers in `src`.
+static std::unordered_set<irep_idt>
+find_symbol_identifiers(const goto_programt &src)
+{
+  std::unordered_set<irep_idt> identifiers;
+  forall_goto_program_instructions(i_it, src)
+  {
+    // compute forward edges first
+    switch(i_it->type())
+    {
+    case ASSERT:
+    case ASSUME:
+    case GOTO:
+      append_ids_in_expr(i_it->condition(), identifiers);
+      break;
+
+    case FUNCTION_CALL:
+      append_ids_in_expr(i_it->call_lhs(), identifiers);
+      for(const auto &e : i_it->call_arguments())
+        append_ids_in_expr(e, identifiers);
+      break;
+    case ASSIGN:
+    case OTHER:
+
+    case SET_RETURN_VALUE:
+    case DECL:
+    case DEAD:
+      for(const auto &e : i_it->code().operands())
+      {
+        append_ids_in_expr(e, identifiers);
+      }
```

Comment:
It should be safe to do `find_symbols(i_it->code(), identifiers);` as that will also iterate over the operands.